consensus/state.go

```diff
@@ -1112,23 +1121,24 @@ func (cs *State) defaultDecideProposal(height int64, round int32) {
 		cs.Logger.Error("enterPropose: Error signing proposal", "height", height, "round", round, "err", err)
 	}
 
-	// TODO(ismail): capture this in the Consensus ADR
-	// post data to ipfs
-	// TODO(evan): don't hard code context and timeout
-	//
-	// longer timeouts result in block proposers failing to propose blocks in time.
-	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*1500)
-	defer cancel()
-	cs.Logger.Info("Putting Block to ipfs", "height", block.Height)
-	// TODO: post data to IPFS in a goroutine
-	err = ipld.PutBlock(ctx, cs.dag, block)
-	if err != nil {
-		// If PutBlock fails we will be the only node that has the data
-		// this means something is seriously wrong and we can not recover
-		// from that automatically.
-		panic(fmt.Sprintf("failure to post block data to IPFS: %s", err.Error()))
+	// cancel ctx for previous proposal block to ensure block putting/providing does not queues up
+	if cs.proposalCancel != nil {
+		cs.proposalCancel()
 	}
-	cs.Logger.Info("Finished putting block to ipfs", "height", block.Height)
+	cs.proposalCtx, cs.proposalCancel = context.WithCancel(context.TODO())
+	go func() {
+		cs.Logger.Info("Putting Block to IPFS", "height", block.Height)
+		err = ipld.PutBlock(cs.proposalCtx, cs.dag, block, cs.croute, cs.Logger)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				cs.Logger.Error("Putting Block didn't finish in time and was terminated", "height", block.Height)
+				return
+			}
+			cs.Logger.Error("Failed to put Block to IPFS", "err", err, "height", block.Height)
+			return
+		}
+		cs.Logger.Info("Finished putting block to IPFS", "height", block.Height)
+	}()
```

Comment:
> but it's a big change.

Why??? Previously that was just a block-saving function

Reply:
@evan-forbes, I guess you miss the thing that now `PutBlock` can take ___minutes___ due to blocking providing what can freeze consensus. 

Btw, I hope you understand why we need that context logic.

Reply:
I should have been clearer, that's my mistake. We've discussed posting proposal block data to IPFS sync and async a few times previously, so I'm bringing to everyone's attention that this PR changes that. I understand that it is necessary, I too ran the experiments to find out exactly how long it takes. However, this detail is not in the description of this PR or documented anywhere else, so I just wanted to make sure that everyone is on the same page. :slightly_smiling_face: 


> Why??? Previously that was just a block-saving function

We currently [panic](https://github.com/lazyledger/lazyledger-core/blob/eab0f8f9c19fc8edbcaa0a6d2a1f1cd19f8a47f1/consensus/state.go#L1125) if something goes wrong, and obvi don't move on with the rest of the proposal process until we are sure that the block is posted to the embedded IPFS node. While entirely necessary, this is a change in consensus, which is inherently a big deal.

Reply:
> condescending

Thanks for the new word I didn't know before. Translation tells me that you are right if we just consider selected code in the GH comment, bc there is a code comment which explains the idea behind the context logic and it's not part of selected GH comments. Although. asking like "Do you understand" would sound much softer and friendlier

Reply:
> Hence, you've changed this to non-blocking. And this intentional and good but it should be clearly communicated why to the reviewer, and more so, in the code itself.

I thought that my investigation comment on the timeout issue and multiple sync discussions explained that well 😞

Reply:
> So at least in its current form, it's actually not such a big deal as validators still reply on the tendermint block gossiping anyways. IMO, we can merge this as is. But should carefully consider this when we actually replace the current block gossiping mechanism.

I thought this through well enough along with experiments to convince myself that's the best way to go. If you go through the commit history, you will see that I changed that part at least 3 times in the code. That should also work well enough, in case we start gossiping in PurBlock. In some way, I am even kinda proud of this code, specifically for how many cases it covers and how it handles context. Currently, I think that's a final state for that unless we decide to take another very different way to work with IPFS related components, which is very unlikely.